axelrod/strategies/axelrod_second.py

```diff
@@ -199,3 +202,214 @@ def strategy(self, opponent: Player) -> Action:
+        two_turns_after_good_defection_ratio and the probability values, 
+        as well as sets the value of num_turns_after_good_defection.
+        """
+        if opponent.history[-1] == D: 
+            self.opponent_consecutive_defections += 1
+        else:
+            self.opponent_consecutive_defections = 0
+
+        if self.num_turns_after_good_defection == 2:
+            self.num_turns_after_good_defection = 0
+            self.two_turns_after_good_defection_ratio = (
+                ((self.two_turns_after_good_defection_ratio 
+                * self.two_turns_after_good_defection_ratio_count) 
+                + (3 - (3 * self.dict[opponent.history[-1]])) 
+                + (2 * self.dict[self.history[-1]]) 
+                - ((self.dict[opponent.history[-1]] 
+                * self.dict[self.history[-1]]))) 
+                / (self.two_turns_after_good_defection_ratio_count + 1)
+                )
+            self.two_turns_after_good_defection_ratio_count += 1
+        elif self.num_turns_after_good_defection == 1:
+            self.num_turns_after_good_defection = 2
+            self.one_turn_after_good_defection_ratio = (
+                ((self.one_turn_after_good_defection_ratio 
+                * self.one_turn_after_good_defection_ratio_count)
+                + (3 - (3 * self.dict[opponent.history[-1]])) 
+                + (2 * self.dict[self.history[-1]]) 
+                - (self.dict[opponent.history[-1]] 
+                * self.dict[self.history[-1]])) 
+                / (self.one_turn_after_good_defection_ratio_count + 1)
+                )
+            self.one_turn_after_good_defection_ratio_count += 1
+   
+    def strategy(self, opponent: Player) -> Action:
+
+        if len(self.history) == 0:
+            return C
+
+       
+        self.update_state(opponent)
+        if  self.num_turns_after_good_defection in [1, 2]:
+            return C        
+
+        current_score = compute_final_score(zip(self.history, opponent.history))
+
+        if (current_score[0] / ((len(self.history)) + 1)) >= 2.25:
+            probability = (
+                (.95 - (((self.one_turn_after_good_defection_ratio)
+                + (self.two_turns_after_good_defection_ratio) - 5) / 15)) 
+                + (1 / (((len(self.history))+1) ** 2))
+                - (self.dict[opponent.history[-1]] / 4)
+                )
+            if random.random() <= probability: 
+                return C
+            self.num_turns_after_good_defection = 1
+            return D
+        elif (current_score[0] / ((len(self.history)) + 1)) >= 1.75:
+            probability = (
+                (.25 + ((opponent.cooperations + 1) / ((len(self.history)) + 1)))
+                - (self.opponent_consecutive_defections * .25) 
+                + ((current_score[0] 
+                - current_score[1]) / 100) 
+                + (4 / ((len(self.history)) + 1))
+                )
+            if random.random() <= probability:
```

Comment:
A note for whoever reviews this after me: we could be tempted to use `random_choice` but `probability` is not really a well defined probability (it can be greater than 1). I suggest we leave it as it is for the sake of simplicity of the code.